### ISSUE TITLE
Rails 6.1 upgrade

### ISSUE THREAD

**Issue by mtk72** [rails-sqlserver/activerecord-sqlserver-adapter#849]
## Issue
Unable to upgrade to Rails 6.1.0.rc1 (full release is about be released and I suspect that it is the same situation)
>bundle update
Fetching gem metadata from https://rubygems.org/............
Fetching gem metadata from https://rubygems.org/.
Resolving dependencies...
Bundler could not find compatible versions for gem "activerecord":
  In Gemfile:
    activerecord-sqlserver-adapter (~> 6.0) was resolved to 6.0.0, which depends on
      activerecord (~> 6.0.0)
    rails (~> 6.1.0.rc1) was resolved to 6.1.0.rc1, which depends on
      activerecord (= 6.1.0.rc1)

## Expected behavior
bundle update to succeed.

## Actual behavior
bundle update fails with error message above.

## How to reproduce
Change Gem for Rails as follows:
gem 'rails', '~> 6.1.0.rc1'
>bundle update

## Details

- **Rails version**: `6.1.0.rc2`
- **SQL Server adapter version**: `6.0.0`

  ```
  run `tsql -C` and paste here the output.
  ```
Compile-time settings (established with the "configure" script)
                            Version: freetds v1.1.6
             freetds.conf directory: /etc/freetds
     MS db-lib source compatibility: no
        Sybase binary compatibility: yes
                      Thread safety: yes
                      iconv library: yes
                        TDS version: auto
                              iODBC: no
                           unixodbc: yes
              SSPI "trusted" logins: no
                           Kerberos: yes
                            OpenSSL: no
                             GnuTLS: yes
                               MARS: yes

**Comment by wpolicarpo** [comment]
Hey, no work has been done for Rails 6.1 yet.

I'm a bit busy at the moment., but I will try to start something in the next few days.

**Comment by marciojg** [comment]
Hy @wpolicarpo any news about this issue?

**Comment by lk0001** [comment]
@marciojg I haven't done a lot of testing, but you can try https://github.com/rails-sqlserver/activerecord-sqlserver-adapter/compare/master...lk0001:master

**Comment by colmexdev** [comment]
> @marciojg I haven't done a lot of testing, but you can try [master...lk0001:master](https://github.com/rails-sqlserver/activerecord-sqlserver-adapter/compare/master...lk0001:master)

Sorry to bother over here, but how can I get this compare commit working on a Rails app?

Thanks in advance.

**Comment by lk0001** [comment]
Use
```
gem 'activerecord-sqlserver-adapter', github: 'lk0001/activerecord-sqlserver-adapter'
```
in Gemfile instead of 
```
gem 'activerecord-sqlserver-adapter'
```

**Comment by johan-smits** [comment]
@lk0001 I'm testing it also but until now all good :+1:, could you submit a PR so that more people can comment on it if they think it needs changes?

**Comment by lk0001** [comment]
https://github.com/rails-sqlserver/activerecord-sqlserver-adapter/pull/853

**Comment by steveny** [comment]
I've tested this and have one issue.  I'm unsure if it is related to the adapter or active record proper.  If you have a column as type "boolean" and a default, this line throws an exception as there is no unary operator for booleans:

https://github.com/rails/rails/blame/0f43f55be62638cf5102dca73cd2b8e61f2444cf/activerecord/lib/active_record/connection_adapters/column.rb#L94

I'll admit that I'm not familiar with this coding idiom, but it seems dangerous in this case.

fwiw, `@default` was "true" using the sqlserver-adapter and "1" for the mysql adapter.

You can repeat by creating a table like:
```
    create_table :categories do |t|
      t.boolean :visible, default: true
    end
```

I monkey patched TrueClass and everything else in our app worked.

**Comment by alastair-measures** [comment]
> I've tested this and have one issue. I'm unsure if it is related to the adapter or active record proper. If you have a column as type "boolean" and a default, this line throws an exception as there is no unary operator for booleans:

From my experience SQL Server thinks of booleans as smaller integers and this is not plastered over by activerecord or this driver (in it's 6.0.0 edition).  

So set your boolean default to 1 or 0 and things will work but possibly only until you switch back to PostgreSQL.  Comparisons will likewise need to be rephrased. 

Hope that helps

**Comment by steveny** [comment]
@alastair-measures The issue is core active record uses the unary operator to freeze strings - for those it freezes, it assumes every attribute used in `deduplicated` is a string.  If you change the default from a boolean to an integer, it is even worse.  It will set the default to be the negative of that integer.  My adapter change above does need to be a bit more clever in that it needs to preserve nil.

**Comment by alastair-measures** [comment]
**@steveny** Thank you for the response.  My experience of SQLServer  booleans did not involve a default in a migration.  It did however involve a permission flag being used in a pundit policy.  On discovering that SQLServer regards the boolean data type as a smaller integer I coded accordingly, it worked and I moved on.  
C was like this in the 1970's and 1980's and surely things could have moved on since then.  ActiveRecord is great but sooner or later one ends up doing DBMS specific code, so one just has to get over it and crack on!

**Comment by marciojg** [comment]
@steveny I agree with you! The issue be in the rails/activerecord.

This is the return of my tests when has `default: true`.

```
** Invoke db:_dump (first_time)
** Execute db:_dump
** Invoke db:schema:dump (first_time)
** Invoke db:load_config 
** Execute db:schema:dump
  SQL (7.6ms)  USE [MSTest]
   (7.6ms)  SELECT [schema_migrations].[version] FROM [schema_migrations] ORDER BY [schema_migrations].[version] ASC
   (7.4ms)  SELECT DATABASEPROPERTYEX(DB_NAME(), 'Collation')
rails aborted!
NoMethodError: undefined method `-@' for true:TrueClass
/usr/local/bundle/gems/activerecord-6.1.2.1/lib/active_record/connection_adapters/column.rb:94:in `deduplicated'
/usr/local/bundle/gems/activerecord-6.1.2.1/lib/active_record/connection_adapters/deduplicable.rb:19:in `deduplicate'
/usr/local/bundle/gems/activerecord-6.1.2.1/lib/active_record/connection_adapters/deduplicable.rb:14:in `new'
/usr/local/bundle/bundler/gems/activerecord-sqlserver-adapter-d88952e16c50/lib/active_record/connection_adapters/sqlserver/schema_statements.rb:87:in `new_column'
/usr/local/bundle/bundler/gems/activerecord-sqlserver-adapter-d88952e16c50/lib/active_record/connection_adapters/sqlserver/schema_statements.rb:73:in `block in columns'
...
```

You consider open a issue in the [rails/activerecord](https://github.com/rails/rails/issues?q=is%3Aopen+is%3Aissue+label%3Aactiverecord) about this?

**Comment by Circuit8** [comment]
I also ran into this issue, and had a go at fixing it here: #854 

Tests pass and the behaviour seems to be in line with what the postgres adapter is doing.

**Comment by joshleblanc** [comment]
It looks like rails 6.1 broke queries with arrays

This change: https://github.com/rails/rails/commit/72fd0bae5948c1169411941aeea6fef4c58f34a9#diff-587b7b18d0042b98a933a40411b7127eff43078ee26cdc0a291f9dafc641a188

Seems to cause the binds to be passed just as strings/numbers rather than the attribute bind. This causes [this line](https://github.com/rails-sqlserver/activerecord-sqlserver-adapter/blob/master/lib/active_record/connection_adapters/sqlserver/database_statements.rb#L319) to fail due to not responding to `type`, as well as [this line](https://github.com/rails-sqlserver/activerecord-sqlserver-adapter/blob/master/lib/active_record/connection_adapters/sqlserver/database_statements.rb#L330) for similar reasons.

**Comment by joshleblanc** [comment]
I've been working off of @Circuit8's branch to fix problems as I come across them. I've added a monkey patch to revert the optimisations in rails 6.1, but I'm not confident that's acceptable for a PR. 

https://github.com/rails-sqlserver/activerecord-sqlserver-adapter/compare/master...joshleblanc:rails-6-1-support

**Comment by wpolicarpo** [comment]
A follow-up for those waiting for a 6.1 release:

There are a [few 6.0 issues](https://github.com/rails-sqlserver/activerecord-sqlserver-adapter/labels/rails-6.0) I want to address before starting a 6.1 release. I don't want to release 6.1 knowing that 6.0 is currently broken but I'm unable to find time to look into them in the upcoming days/weeks, so all PRs are more than welcome.

That list is not extensive as I also have plans to move all the CI configuration to Github Actions and change the versioning scheme for future releases, but that's not an impediment for 6.1, just nice to have.

**Comment by runephilosof-abtion** [comment]
The only open 6.0 issue is missing `upsert_all` support.
I don't think it is important to have that implemented before supporting 6.1, which we already have a working PR for: 
https://github.com/rails-sqlserver/activerecord-sqlserver-adapter/pull/861

**Comment by wpolicarpo** [comment]
Well, I wouldn't say it's not important (maybe it's not for you @runephilosof-abtion), but I kind of agree we can move on and open a 6.1 WIP release.

I moved all the CI stuff to GitHub Actions and added 3.0.x to the matrix, which revealed a failing test addressed by #868. We are now investigating why we need to coerce that test and why the SQL Server adapter does not implement `DecimalWithoutScale`. If nothing arises in the coming days, I'll merge that PR and cut a new 6.0.x release and move it to a stable branch.

**Comment by ecentell-CPF** [comment]
@wpolicarpo FYI:
The version 6.0.2 you pushed to Rubygems.org is not working for 6.1.  

gem 'activerecord-sqlserver-adapter', '~> 6.0', '>= 6.0.2' 
Bundler could not find compatible versions for gem "activerecord":
  In Gemfile:
activerecord-sqlserver-adapter (~> 6.0, >= 6.0.2) was resolved to 6.0.2,
which depends on
      activerecord (~> 6.0.0)

If I change to 
gem 'activerecord-sqlserver-adapter', github: 'rails-sqlserver/activerecord-sqlserver-adapter', ref: '1f46d2c062f8ed172468d52eb28caa602b6578b9' 

I can bundle. But then receive the following error on running the rails server
C:/Ruby27-x64/lib/ruby/gems/2.7.0/bundler/gems/activerecord-sqlserver-adapter-1f46d2c062f8/lib/active_record/connection_adapters/sqlserver/schema_creation.rb:6:in `<module:SQLServer>': uninitialized constant ActiveRecord::ConnectionAdapters::AbstractAdapter::SchemaCreation (NameError)
Did you mean?  ActiveRecord::ConnectionAdapters::SchemaCreation
               ActiveRecord::SchemaMigration

If I change to the testing version from lk0001 above
gem 'activerecord-sqlserver-adapter', github: 'lk0001/activerecord-sqlserver-adapter'
I can run my rails server.

Please advise.

**Comment by wpolicarpo** [comment]
@ecentell-CPF Rails 6.1 support is still WIP and I don't recommend you to use that branch simply because it's broken.

I suggest you to have a look at the [README](https://github.com/rails-sqlserver/activerecord-sqlserver-adapter#about-the-adapter) where it explains the rationale behind the adapter versioning. 6.0.x is compatible with Rails 6.0.x and 6.1.x will be compatible with Rails 6.1.x and so on.

**Comment by wpolicarpo** [comment]
I've just released [6.1.0.0.rc1](https://rubygems.org/gems/activerecord-sqlserver-adapter/versions/6.1.0.0.rc1).

Please, report any bugs you may find. If nothing major is reported, I'll cut the final 6.1.0.0 release in 2 weeks.

And please, do no report bugs in this thread. Open a new issue so it's easier to track changes.

**Comment by wpolicarpo** [comment]
Forgot to thanks everyone who helped with PRs and bug reports. Special thanks to @aidanharan and @mgrunberg who did most of the work for this release to happen.

**Comment by mgrunberg** [comment]
@wpolicarpo is it safe to release version 6.1.0.0? Sorry if this is not the proper place to ask this.

**Comment by wpolicarpo** [comment]
@mgrunberg sure thing. I'll cut the final release tomorrow

**Comment by wpolicarpo** [comment]
Final 6.1.0.0 was released, sorry for the delay.
